### PR TITLE
str.istitle() support for unicode string

### PR DESCRIPTION
**Pull request by rdesai16**
This PR enables support for istitle method for unicode strings

### CODE REVIEW

**Comment by pearu** [comment]
I am wondering if the used approach of re-implementing `istitle` method (and potentially all other `str` methods) within numba is reasonable?

Another approach would be to convert numba `UnicodeType` to Python `str` and re-use `str` methods implementations. For instance, in PR #4425 the `str` is overloaded for `UnicodeCharSeq`, see 
https://github.com/numba/numba/pull/4425/files#diff-8b7751de7f31eab2823b0beb825e8c66R209 and then implementing `hash` for `UnicdeCharSeq` became trivial:
```
@overload_method(types.UnicodeCharSeq, '__hash__')
def unicode_charseq_hash(s):
    def impl(s):
        return hash(str(s))
    return impl
```
Similarly, one could overload `str` for `UnicodeType` and then re-use `str.istitle`:
```
@overload_method(types.UnicodeType, 'istitle')
def unicode_istitle(s):
    def impl(s):
        return str(s).istitle()
    return impl
```
This approach would allow quickly exposing all `str` methods to `UnicodeType` (with an overhead of `UnicodeType` -> `str` transformation, of course).

**Comment by stuartarchibald** [comment]
> I am wondering if the used approach of re-implementing `istitle` method (and potentially all other `str` methods) within numba is reasonable?
> 
> Another approach would be to convert numba `UnicodeType` to Python `str` and re-use `str` methods implementations. For instance, in PR #4425 the `str` is overloaded for `UnicodeCharSeq`, see
> https://github.com/numba/numba/pull/4425/files#diff-8b7751de7f31eab2823b0beb825e8c66R209 and then implementing `hash` for `UnicdeCharSeq` became trivial:
> 

@pearu I think that fundamentally implementations of these methods have to be written as they do not exist anywhere in Numba. It seems like what you are describing is a method of making is easier to reuse the implementations in the case that they existed?

--------------------------------
> ```python
> @overload_method(types.UnicodeCharSeq, '__hash__')
> def unicode_charseq_hash(s):
>     def impl(s):
>         return hash(str(s))
>     return impl
> ```

This only worked because `__hash__` for the `UnicodeType` was already implemented https://github.com/numba/numba/blob/4ab57e65d3ab3d44207f1c1acec5b9389ab964c1/numba/targets/hashing.py#L645-L661 and `hash` has been implemented to call `<instance>.__hash__()` https://github.com/numba/numba/blob/4ab57e65d3ab3d44207f1c1acec5b9389ab964c1/numba/targets/hashing.py#L61-L65 The `str` overload here https://github.com/numba/numba/pull/4425/files#diff-8b7751de7f31eab2823b0beb825e8c66R307-R321 is converting the `UnicodeCharSeq` to Numba's internal unicode string representation via `unicode._empty_string` which is typed to return a `types.UnicodeType` https://github.com/numba/numba/blob/4ab57e65d3ab3d44207f1c1acec5b9389ab964c1/numba/unicode.py#L210-L246 with code generated around the `UnicodeModel` data representation https://github.com/numba/numba/blob/4ab57e65d3ab3d44207f1c1acec5b9389ab964c1/numba/unicode.py#L42-L55 Net result is that `types.UnicodeCharSeq` was converted to `types.UnicodeType` and the implementation detail for the rest of the call sequence was already working, hence the above worked.
 
-------------------

> Similarly, one could overload `str` for `UnicodeType` and then re-use `str.istitle`:
> 
> ```python
> @overload_method(types.UnicodeType, 'istitle')
> def unicode_istitle(s):
>     def impl(s):
>         return str(s).istitle()
>     return impl
> ```
> 
> This approach would allow quickly exposing all `str` methods to `UnicodeType` (with an overhead of `UnicodeType` -> `str` transformation, of course).

As noted above, I think this will work and does make good reuse of code, but requires implementing the functionality on `type.Unicode` first.

**Comment by stuartarchibald** [comment]
WRT all these `isFOO()` methods, I'm wondering if it's possible to just codegen the lot, looks like it's a repeat pattern.

**Comment by pearu** [comment]
yes, @stuartarchibald , you are right: the njitted functions use `UnicodeType` implementations when calling `str(s).isFOO()`.

Would it be possible to call the corresponding implementations via Python C/API?

**Comment by stuartarchibald** [comment]
> Would it be possible to call the corresponding implementations via Python C/API?

Yes, but it the GIL would have to be reacquired which is not desirable. In cases where the complexity of reimplementing the CPython function is too high e.g. `str.format` this may be done, but in general reimplementation in `@overload` or `@intrinsic` or copying and adapting the CPython source into the Numba code base and calling that is preferred.

File: numba/unicode.py [hunk 1]
```diff
@@ -1148,6 +1148,56 @@ def impl(a):
     return impl
 
 
+def _is_title(is_lower, is_upper, is_title):
+    # This algorithm is an amalgamation of:
+    # https://github.com/python/cpython/blob/master/Objects/unicodeobject.c#L11902-L11948
+    # https://github.com/python/cpython/blob/master/Objects/bytes_methods.c#L254-L291
+    def impl(a):
```
Comment (stuartarchibald) [inline numba/unicode.py]: I'm wondering whether all these `isFOO` methods could be generated? The code structure of the algorithm seems pretty much identical, so much that there's already an element of this here: https://github.com/numba/numba/pull/4442/files#diff-a21787256ad779b4c4c5c99915c71025R1181-R1185? Thoughts?

**Comment by stuartarchibald** [comment]
Thanks for the PR, I'm of the view that we should consider https://github.com/numba/numba/pull/4442#discussion_r314353276 as a concept/idea ahead of formal code review.

**Comment by stuartarchibald** [comment]
Closing as #4645 continued the implementation and was merged.

**Pull request closed without merging**